src/builder/QuarkBuilderBase.sol

```diff
@@ -0,0 +1,642 @@
+                paymentTokenInAssetSymbolOut = true;
+            }
+
+            if (
+                needsBridgedFunds(
+                    actionIntent.assetSymbolOuts[i],
+                    actionIntent.amountOuts[i],
+                    actionIntent.chainId,
+                    chainAccountsList,
+                    payment
+                )
+            ) {
+                if (actionIntent.bridgeEnabled) {
+                    uint256 amountNeededOnDst = actionIntent.amountOuts[i];
+                    if (
+                        payment.isToken && Strings.stringEqIgnoreCase(payment.currency, actionIntent.assetSymbolOuts[i])
+                    ) {
+                        amountNeededOnDst += PaymentInfo.findMaxCost(payment, actionIntent.chainId);
+                    }
+
+                    (IQuarkWallet.QuarkOperation[] memory bridgeQuarkOperations, Actions.Action[] memory bridgeActions)
+                    = Actions.constructBridgeOperations(
+                        Actions.BridgeOperationInfo({
+                            assetSymbol: actionIntent.assetSymbolOuts[i],
+                            amountNeededOnDst: amountNeededOnDst,
+                            dstChainId: actionIntent.chainId,
+                            recipient: actionIntent.actor,
+                            blockTimestamp: actionIntent.blockTimestamp,
+                            useQuotecall: actionIntent.useQuotecall
+                        }),
+                        chainAccountsList,
+                        payment
+                    );
+
+                    for (uint256 j = 0; j < bridgeQuarkOperations.length; ++j) {
+                        List.addQuarkOperation(quarkOperations, bridgeQuarkOperations[j]);
+                        List.addAction(actions, bridgeActions[j]);
+                    }
+                } else {
+                    uint256 balanceOnChain = getBalanceOnChain(
+                        actionIntent.assetSymbolOuts[i], actionIntent.chainId, chainAccountsList, payment
+                    );
+                    uint256 amountNeededOnChain = getAmountNeededOnChain(
+                        actionIntent.assetSymbolOuts[i], actionIntent.amountOuts[i], actionIntent.chainId, payment
+                    );
+                    uint256 maxCostOnChain =
+                        payment.isToken ? PaymentInfo.findMaxCost(payment, actionIntent.chainId) : 0;
+                    uint256 availableAssetBalance =
+                        balanceOnChain >= maxCostOnChain ? balanceOnChain - maxCostOnChain : 0;
+                    revert FundsUnavailable(actionIntent.assetSymbolOuts[i], amountNeededOnChain, availableAssetBalance);
+                }
+            }
+        }
+
+        if (payment.isToken && !paymentTokenInAssetSymbolOut) {
+            uint256 maxCostOnDstChain = PaymentInfo.findMaxCost(payment, actionIntent.chainId);
+
+            for (uint256 k = 0; k < actionIntent.assetSymbolIns.length; ++k) {
```

Comment:
Some documentation on these pieces of logic would be nice. Especially since documentation used to exist on a lot of the logic and are now lost.